Mathlib/RingTheory/Regular/ProjectiveDimension.lean

```diff
@@ -0,0 +1,334 @@
+    · have : N = ⊤ := (WithBot.coe_unbot _ eqbot).symm.trans (WithBot.coe_inj.mpr eqtop)
+      simp [this]
+    · let n := (N.unbot eqbot).toNat
+      have : N = n := (WithBot.coe_unbot _ eqbot).symm.trans
+        (WithBot.coe_inj.mpr (ENat.coe_toNat eqtop).symm)
+      simpa only [this] using aux n
+
+lemma projectiveDimension_quotient_regular_sequence [Small.{v} R] (M : ModuleCat.{v} R)
+    [Nontrivial M] [Module.Finite R M] (rs : List R) (reg : IsWeaklyRegular M rs)
+    (mem : ∀ r ∈ rs, r ∈ maximalIdeal R) :
+    projectiveDimension (ModuleCat.of R (M ⧸ Ideal.ofList rs • (⊤ : Submodule R M))) =
+    projectiveDimension M + rs.length := by
+  generalize len : rs.length = n
+  induction n generalizing M rs
+  · rw [List.length_eq_zero_iff.mp len, Ideal.ofList_nil, Submodule.bot_smul]
+    simpa using projectiveDimension_eq_of_iso (Submodule.quotEquivOfEqBot ⊥ rfl).toModuleIso
+  · rename_i n hn _ _
+    match rs with
+    | [] => simp at len
+    | x :: rs' =>
+      simp only [List.mem_cons, forall_eq_or_imp] at mem
+      let _ : Nontrivial (QuotSMulTop x M) := quotSMulTop_nontrivial' mem.1 M
+      simp only [Nat.cast_add, Nat.cast_one]
+      simp only [List.length_cons, Nat.add_right_cancel_iff] at len
+      have : IsSMulRegular M x := ((isWeaklyRegular_cons_iff M _ _).mp reg).1
+      rw [projectiveDimension_eq_of_iso
+        (Submodule.quotOfListConsSMulTopEquivQuotSMulTopInner M x rs').toModuleIso, add_comm _ 1,
+        ← add_assoc, ← projectiveDimension_quotSMulTop_eq_succ_of_isSMulRegular M x this mem.1,
+        ← hn (ModuleCat.of R (QuotSMulTop x M)) rs' ((isWeaklyRegular_cons_iff M _ _).mp reg).2
+          mem.2 len]
+
+omit [IsLocalRing R] [IsNoetherianRing R] in
+lemma projectiveDimension_eq_zero_of_projective (M : ModuleCat.{v} R) [Projective M]
+    [Nontrivial M] : projectiveDimension M = 0 := by
+  apply le_antisymm
+  · rw [← Nat.cast_zero, projectiveDimension_le_iff M 0]
+    infer_instance
+  · rw [← Nat.cast_zero, projectiveDimension_ge_iff M 0, hasProjectiveDimensionLT_zero_iff_isZero,
+      ModuleCat.isZero_iff_subsingleton, not_subsingleton_iff_nontrivial]
+    assumption
+
+section
+
+variable {M N : Type*} [AddCommGroup M] [Module R M] [AddCommGroup N] [Module R N]
+
+/-- The linear map `M/IM → N/IN` induced by `M → N`. -/
+def Ideal.smulTopLinearMap (I : Ideal R) (f : M →ₗ[R] N) :
+    M ⧸ I • (⊤ : Submodule R M) →ₗ[R] N ⧸ I • (⊤ : Submodule R N) :=
+  Submodule.mapQ _ _ f (Submodule.smul_top_le_comap_smul_top I f)
+
+/-- The linear map `M/IM ≃ N/IN` induced by `M ≃ N`. -/
+def Ideal.smulTopLinearEquiv (I : Ideal R) (e : M ≃ₗ[R] N) :
+    (M ⧸ I • (⊤ : Submodule R M)) ≃ₗ[R] (N ⧸ I • (⊤ : Submodule R N)) where
+  __ := Ideal.smulTopLinearMap I e
+  invFun := Ideal.smulTopLinearMap I e.symm
+  left_inv y := by
+    induction y using Submodule.Quotient.induction_on
+    simp [smulTopLinearMap]
+  right_inv y := by
+    induction y using Submodule.Quotient.induction_on
+    simp [smulTopLinearMap]
```

Comment:
These shouldn't be here.
(And I'm not sure if we need the first one as a separate def)